From 20e75cc0aa5bb478db7070f358339ff66ba3bf2e Mon Sep 17 00:00:00 2001
From: Mike Blumenkrantz <michael.blumenkrantz@gmail.com>
Date: Thu, 19 May 2022 19:03:03 -0400
Subject: [PATCH] nir/lower_atomics_to_ssbo: set ACCESS_VOLATILE on atomic ops

these ops aren't regular intrinsics, they're atomic ops, so they must
have the corresponding decoration to avoid having optimizations combine
loads/stores (e.g., nir_opt_copy_prop_vars)

cc: mesa-stable

fixes:
dEQP-GLES31.functional.atomic_counter*
---
 src/compiler/nir/nir_lower_atomics_to_ssbo.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/compiler/nir/nir_lower_atomics_to_ssbo.c b/src/compiler/nir/nir_lower_atomics_to_ssbo.c
index 448f63bdc7cc..0ead10bf11e7 100644
--- a/src/compiler/nir/nir_lower_atomics_to_ssbo.c
+++ b/src/compiler/nir/nir_lower_atomics_to_ssbo.c
@@ -132,6 +132,7 @@ lower_instr(nir_intrinsic_instr *instr, unsigned ssbo_offset, nir_builder *b)
        */
       new_instr->num_components = instr->dest.ssa.num_components;
    }
+   nir_intrinsic_set_access(new_instr, ACCESS_VOLATILE);
 
    nir_ssa_dest_init(&new_instr->instr, &new_instr->dest,
                      instr->dest.ssa.num_components,
-- 
GitLab

